### ISSUE TITLE
Forced to clean build after code changes

### ISSUE THREAD

**Issue by jorgenfaunaapp** [google/dagger#2799]
Hi, I am having some problems with Hilt dependency injection. If I clean the project and build, it builds just fine. But gives me ‘Unresolved reference’ on a bunch of my injected view models if I change anything in my activities/fragments and try and build again without cleaning. Some builds also trigger a circular dependency error, but I cannot figure out where the should be occuring.

I have tried all different kinds of configurations in my gradle files, to no avail, I have tried to paste my configuration with the parts concerning dagger and gradle below:

**build.gradle file**

```
buildscript {
    ext.kotlin_version = '1.5.10'
    ext.hilt_version = '2.37'

    repositories {
        google()
        jcenter()

    }
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "com.google.dagger:hilt-android-gradle-plugin:$hilt_version"
    }
}
```

**build.gradle app file**

```
plugins {
    id("dagger.hilt.android.plugin")
}

dependencies {
    implementation "com.google.dagger:hilt-android:$hilt_version"
    kapt "com.google.dagger:hilt-android-compiler:$hilt_version"
}
```

I get the following error messages on subsequent builds, either unresolved reference or dependency cycle

<img width="831" alt="Screenshot 2021-07-07 at 11 31 59" src="https://user-images.githubusercontent.com/69505339/128153898-a8ac3d50-28e1-4566-b2d3-2fd93a6a4479.png">

<img width="1585" alt="Screenshot 2021-07-07 at 11 58 28" src="https://user-images.githubusercontent.com/69505339/128154060-7fe81761-ad1e-4968-8d0e-71a562855bdb.png">

Any help would be appreciated. I have exhausted a lot of different options by now.

**Comment by jorgenfaunaapp** [comment]
Yeah, no I have tried that. Still gives me Unresolved reference, all over the place.

**Comment by bcorso** [comment]
@jorgenfaunaapp just to confirm, after enabling correctErrorTypes are you still seeing the Dagger/DependencyCycle error? If so can you post the updated stacktrace without the NonExistentClass?

For the unresolved references, I don't see anything relating it to Hilt. Do you have a reason to suspect that this is related to Hilt?

**Comment by jorgenfaunaapp** [comment]
I cannot reproduce the circular dependency as of now, so maybe they are gone for good. Sorry for not writing that in my previous comment. I will see if it comes up again when I am back working with the project, they seemed to come at random between the Unresolved reference ones.

The reason why I expect it to be something to do with hilt is that it is my dependency injected classes that it does not understand the reference too. Sorry if not, but that just seemed like the most natural place to look, and where all my googling has lead me.

I have the following in my MainActivity

```
import com.fauna.fauna.vaccinationCards.VaccinationCardsViewModel

@AndroidEntryPoint
class MainActivity : AppCompatActivity(), CoroutineScope {

      private val vaccinationCardsViewModel: VaccinationCardsViewModel by viewModels()

}
```

I have the VaccinationCardsViewModel defined as the following

```
@HiltViewModel
class VaccinationCardsViewModel @Inject constructor(
    private val vaccinationCardsRepository: VaccinationCardsRepository,
    private val notificationService: NotificationService
) : SnackBarViewModel() {

}
```

The updated stack trace with unresolved reference looks like this
<img width="1777" alt="Screenshot 2021-08-06 at 09 08 18" src="https://user-images.githubusercontent.com/69505339/128470711-63bb4993-8c16-4f0d-bc6b-08434a78c615.png">

If you have any ideas to what it could be it would be a massive help.

**Comment by bcorso** [comment]
Which version of Hilt are you on?

Are `MainActivity` and `VaccinationCardsViewModel` in the same Gradle module or separate ones?

If they are in separate Gradle modules, can you check that the `build.gradle` file for `MainActivity` has an explicit `implementation` dependency on the Gradle module containing `VaccinationCardsViewModel`?

Also, do you happen to have [`enableAggregatingTask = true`](https://dagger.dev/hilt/gradle-setup#aggregating-task)?

If so, can you try replacing it with `enableExperimentalClasspathAggregation = true` and see if the issue goes away.

**Comment by jorgenfaunaapp** [comment]
I am currently using hilt version 2.37.
I only have one Gradle module.

I do not have ` hilt { enableAggregatingTask = true }`. Adding `hilt { enableExperimentalClasspathAggregation = true } `did not work.

**Comment by bcorso** [comment]
Since the unresolved references, like `VacinationCardViewModel`, are in the same source set as `MainActivity`, it seems very weird that they can't be resolved from `MainActivity`. The only thing I can think of is that you've forgotten the import statements for those classes.

At the moment, it doesn't sound like this is related to Hilt so I'm going to close this bug. Feel free to reopen if you have more information that suggests it may be related to Hilt.

**Comment by jorgenfaunaapp** [comment]
It is the import statements of the VaccinationCardViewModel that 'trigger' the Unresolved reference.

**Comment by jorgenfaunaapp** [comment]
I finally solved it!  I am also using apollo-android to generated my types, mutations and queries for usage within my android project. In the Apollo schema directory I had a subfolder called `User`, I also had a package in my android project called `user`. This leads to two separate generated folders upon compilation, which for some unbeknownst reason works in the initial builds (or if changes are applied to all files in the package with identical naming) but not in subsequent builds.

I have created a question on the apollo-android repo, trying to explain the error in more detail. I do not know where to go from here, but I just wanted to share the problem in case anyone else had this very undescriptive error message.

apollo-android question:
https://github.com/apollographql/apollo-android/issues/3306

Best regards Jørgen.
